data/lua/wml-tags.lua

```diff
@@ -1787,3 +1787,110 @@ wml_actions.teleport = function(cfg)
 	end
 	wesnoth.teleport(unit, cfg.check_passability == false, cfg.clear_shroud ~= false, cfg.animate)
 end
+
+function wml_actions.set_variable(cfg)
+	local name = cfg.name or helper.wml_error "trying to set a variable with an empty name"
+	local var = wesnoth.get_variable(name)
+
+	if cfg.value ~= nil then -- check for nil because user may try to set a variable as false
+		wesnoth.set_variable(name, cfg.__shallow_parsed.value)
+	elseif cfg.literal ~= nil then
+		wesnoth.set_variable(name, cfg.__shallow_literal.literal)
+	elseif cfg.to_variable then
+		wesnoth.set_variable(name, wesnoth.get_variable(cfg.to_variable))
+	elseif cfg.add then
+		wesnoth.set_variable(name, (tonumber(var) or 0) + (tonumber(cfg.add) or 0))
+	elseif cfg.sub then
+		wesnoth.set_variable(name, (tonumber(var) or 0) - (tonumber(cfg.sub) or 0))
+	elseif cfg.multiply then
+		wesnoth.set_variable(name, (tonumber(var) or 0) * (tonumber(cfg.multiply) or 0))
+	elseif cfg.divide then
+		local divide = tonumber(cfg.divide) or 0
+		if divide == 0 then helper.wml_error("division by zero on variable " .. name) end
+		wesnoth.set_variable(name, (tonumber(var) or 0) / divide)
+	elseif cfg.modulo then
+		local modulo = tonumber(cfg.modulo) or 0
+		if modulo == 0 then helper.wml_error("division by zero on variable " .. name) end
+		wesnoth.set_variable(name, (tonumber(var) or 0) % modulo)
+	elseif cfg.round then
+		local round_val = cfg.round
+		if round_val == "ceil" then
+			wesnoth.set_variable(name, math.ceil(tonumber(var) or 0))
+		elseif round_val == "floor" then
+			wesnoth.set_variable(name, math.floor(tonumber(var) or 0))
+		else
+			local decimals, discarded = math.modf(tonumber(round_val) or 0)
+			local value = (tonumber(var) or 0) * (10 ^ decimals)
+			value = helper.round(value)
+			value = value * (10 ^ -decimals)
+			wesnoth.set_variable(name, value)
+		end
+	-- unlike the other math operations, ipart and fpart do not act on
+	-- the value already contained in the variable
+	-- but on the value assigned to the respective key
+	elseif cfg.ipart then
+		local ivalue, fvalue = math.modf(tonumber(cfg.ipart) or 0)
+		wesnoth.set_variable(name, ivalue)
+	elseif cfg.fpart then
+		local ivalue, fvalue = math.modf(tonumber(cfg.fpart) or 0)
+		wesnoth.set_variable(name, fvalue)
+	elseif cfg.string_length ~= nil then
+		wesnoth.set_variable(name, string.len(tostring(cfg.string_length)))
+	elseif cfg.time then
+		if cfg.time == "stamp" then
+			wesnoth.set_variable(name, wesnoth.get_time_stamp())
+		end
+	elseif cfg.rand then
```

Comment:
As I understand, helper.rand simply calls [set_variable] to leverage its random functionality. If [set_variable] is to be ported to Lua, I think it would make more sense to reverse that – move this code to helper.rand and call it from [set_variable].